File: Labs/week6/C_radix_sort.py
def get_digit(num, d):
    for _ in range(d):
        num //= 10
    return num % 10

def radix_sort(arr, max_digits):
    for d in range(max_digits):
        cnt = [0] * 10
        for x in arr:
            cnt[get_digit(x, d)] += 1
        for i in range(1, 10):
            cnt[i] += cnt[i - 1]
        tmp = [0] * len(arr)
        for i in range(len(arr) - 1, -1, -1):
            digit = get_digit(arr[i], d)
            cnt[digit] -= 1
            tmp[cnt[digit]] = arr[i]
        arr[:] = tmp
        print(' '.join(map(str, arr)))
        if d == 2:
            break

n = int(input())
arr = list(map(int, input().split()))
max_val = max(arr)
max_digits = len(str(max_val)) if max_val > 0 else 1
radix_sort(arr, max_digits)

File: Labs/week6/C_radix_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int getDigit(int num, int d) {
    for (int i = 0; i < d; ++i) num /= 10;
    return num % 10;
}

void radixSort(vector<int>& arr, int maxDigits) {
    vector<int> tmp(arr.size());
    for (int d = 0; d < maxDigits; ++d) {
        int cnt[10] = {0};
        for (int x : arr) cnt[getDigit(x, d)]++;
        for (int i = 1; i < 10; ++i) cnt[i] += cnt[i - 1];
        for (int i = (int)arr.size() - 1; i >= 0; --i) {
            int digit = getDigit(arr[i], d);
            tmp[--cnt[digit]] = arr[i];
        }
        arr = tmp;
        for (size_t i = 0; i < arr.size(); ++i) cout << arr[i] << " ";
        cout << endl;
        if (d == 2) break;
    }
}

int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    int maxVal = 0;
    for (int i = 0; i < n; ++i) {
        cin >> arr[i];
        if (arr[i] > maxVal) maxVal = arr[i];
    }
    int maxDigits = 0;
    while (maxVal > 0) {
        maxDigits++;
        maxVal /= 10;
    }
    radixSort(arr, maxDigits);
    return 0;
}
